apache2/waf_logging/waf_log_util.cc

```diff
@@ -0,0 +1,133 @@
+    waf_format::Properties *properties;
+    waf_format::Details *details;
+
+    properties = waf_format->mutable_properties();
+    details = properties->mutable_details();
+
+    if (resourceId != NULL) {
+        waf_format->set_resourceid(resourceId);
+    }
+
+    if (operationName != NULL) {
+        waf_format->set_operationname(operationName);
+    }
+
+    if (time != NULL) {
+        waf_format->set_time(time);
+    }
+
+    if (category != NULL) {
+        waf_format->set_category(category);
+    }
+
+    if (instanceId != NULL) {
+        properties->set_instanceid(instanceId);
+    }
+
+    if (clientIP != NULL) {
+        properties->set_clientip(clientIP);
+    }
+    
+    if (clientPort != NULL) {
+        properties->set_clientport(clientPort);
+    }
+    
+    if (requestUri != NULL) {
+        properties->set_requesturi(requestUri);
+    }
+    
+    if (ruleSetType != NULL) {
+        properties->set_rulesettype(ruleSetType);
+    }
+    
+    if (ruleSetVersion != NULL) {
+        properties->set_rulesetversion(ruleSetVersion);
+    }
+    
+    if (ruleId != NULL) {
+        properties->set_ruleid(ruleId);
+    }
+    
+    if (messages != NULL) {
+        properties->set_messages(messages);
+    }
+    
+    switch(action) {
+        case 1:
+            properties->set_action(waf_format::Properties::Detected);
+            break;
+        case 2:
+            properties->set_action(waf_format::Properties::Blocked);
+            break;
+        default:
+            break;
+    }
+    
+    if (site == 0) {
+        properties->set_site(waf_format::Properties::Global);
+    }
+    
+    if (details_messages != NULL) {
+        details->set_messages(details_messages);
+    }
+    
+    if (details_data != NULL) {
+        details->set_data(details_data);
+    }
+    
+    if (details_file != NULL) {
+        details->set_file(details_file);
+    }
+    
+    if (details_line != NULL) {
+        details->set_line(details_line);
+    }
+    
+    if (hostname != NULL) {
+        properties->set_hostname(hostname);
+    }
+}
+
+// Main function:  get fields from modsec, set the protobuf object and write to file in json.
+int generate_json(char** result_json, char* resourceId, char* operationName, char* category, char* instanceId, char* clientIP, char* clientPort, char* requestUri, char* ruleSetType, char* ruleSetVersion, char* ruleId, char* messages, int action, int site, char* details_messages, char* details_data, char* details_file, char* details_line, char* hostname, char* time) {
+    waf_format::Waf_Format waf_format;
+    std::string json_string;
+    google::protobuf::util::JsonPrintOptions options;
+    google::protobuf::util::Status convert_result;
+    char* json_str;
+    
+    try {
+        // Verify that the version of the library that we linked against is
+        // compatible with the version of the headers we compiled against.
+        GOOGLE_PROTOBUF_VERIFY_VERSION;
```

Comment:
Should we do this just once during initialization?

Reply:
I think so, but i don't want to expose this to the modsec or add wrapper to make it complicated, also for the performance wise, it doesn't degrade.. so i think it should be fine